burotel/indico_burotel/cli.py

```diff
@@ -0,0 +1,242 @@
+ROOM_FIELDS = ('id', 'division', 'building', 'floor', 'number', 'verbose_name', 'owner', 'acl_entries')
+group_cache = {}
+latlon_cache = {}
+user_cache = {}
+
+
+@cli_group(name='burotel')
+def cli():
+    """Manage the Burotel plugin."""
+
+
+def check_changed_fields(original, new):
+    diff = []
+    for field in ROOM_FIELDS:
+        if field == 'acl_entries':
+            orig_value = {e.principal for e in original.acl_entries}
+            target_value = new['acl_entries']
+        else:
+            orig_value = getattr(original, field)
+            target_value = new[field]
+        if orig_value != target_value:
+            diff.append((field, orig_value, target_value))
+    return diff
+
+
+def get_location(building):
+    location = Location.query.filter(Location.name == "Area {}".format(building)).first()
+    if not location:
+        location = Location(name="Area {}".format(building))
+        print(cformat("%{green!}+%{reset} Adding new location for building {}").format(building))
+        db.session.add(location)
+        db.session.flush()
+    return location
+
+
+def get_user(email):
+    if email not in user_cache:
+        user_cache[email] = get_user_by_email(email)
+    return user_cache[email]
+
+
+def get_principal(name):
+    if '@' in name:
+        return get_user(name)
+
+    # otherwise we assume it's a group's name
+    group = group_cache.setdefault(name, GroupProxy(name, provider='cern-ldap'))
+    if not group or not group.group:
+        group = None
+        print cformat("%{red}!%{reset} Group %{cyan}{}%{reset} doesn't seem to exist!").format(name)
+    return group
+
+
+def get_room(room_id):
+    room = Room.get(room_id)
+    if not room:
+        print cformat("%{yellow}! Desk with ID {} not found.").format(room_id)
+    return room
+
+
+def change_room(room, changes):
+    for field, _, new_value in changes:
+        if field == 'acl_entries':
+            # clear the ACL and add the new principals
+            room.acl_entries.clear()
+            db.session.flush()
+            for p in new_value:
+                room.update_principal(p, full_access=True)
+        else:
+            setattr(room, field, new_value)
+
+
+def _print_changes(room, changes):
+    print '[{}]:'.format(room)
+    for field, old, new in changes:
+        if field == 'acl_entries':
+            old = {e.name for e in old}
+            new = {e.name for e in new}
+        print (cformat(' %{yellow}>%{reset} %{cyan}{}%{reset}: %{red}{}%{reset} -> %{green}{}%{reset}')
+               .format(field, old, new))
+    print
+
+
+def _principal_repr(p):
+    return getattr(p.principal, 'email', p.principal.name)
+
+
+def get_latlon_building(building_num):
+    if building_num not in latlon_cache:
+        # this API request should get the positions of a building's
+        # entrance doors
+        r = requests.get(GIS_URL.format(building_num))
+        data = r.json()
```

Comment:
Why not `data = requests.get(GIS_URL.format(building_num)).json()`?

Reply:
Because I'm an artist. Artists can choose to shape their work the way they want. Stop oppressing me.